File: src/main/python/AoC2024_22.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 22
#

import multiprocessing
import os
import sys
from collections import defaultdict
from collections.abc import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = list[int]
Output1 = int
Output2 = int


TEST1 = """\
1
10
100
2024
"""
TEST2 = """\
1
2
3
2024
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return list(map(int, input_data))

    def secret(self, num: int) -> int:
        num = (num ^ (num * 64)) % 16777216
        num = (num ^ (num // 32)) % 16777216
        return (num ^ (num * 2048)) % 16777216

    def solve(
        self, seeds: Input, worker: Callable[[str, list[int]], None]
    ) -> None:
        if sys.platform.startswith("win"):
            worker("main", seeds)
        else:
            n = os.process_cpu_count()
            batch = [list[int]() for _ in range(n)]
            for cnt, seed in enumerate(seeds):
                batch[cnt % n].append(seed)
            jobs = []
            for i in range(n):
                p = multiprocessing.Process(
                    target=worker, args=(str(i), batch[i])
                )
                jobs.append(p)
                p.start()
            for p in jobs:
                p.join()

    def part_1(self, seeds: Input) -> Output1:
        m_ans = multiprocessing.Manager().dict()

        def worker(pid: str, seeds: list[int]) -> None:
            ans = 0
            for num in seeds:
                n = num
                for _ in range(2000):
                    n = self.secret(n)
                ans += n
            m_ans[pid] = ans

        self.solve(seeds, worker)
        return sum(m_ans.values())

    def part_2(self, seeds: Input) -> Output2:
        m_ans = multiprocessing.Array("i", [0 for _ in range(19**4)])

        def worker(_pid: str, seeds: list[int]) -> None:
            p = defaultdict[int, int](int)
            seen = [-1 for _ in range(19**4)]
            for i, num in enumerate(seeds):
                na = num
                nb = self.secret(na)
                nc = self.secret(nb)
                nd = self.secret(nc)
                b, c, d = (
                    na % 10 - nb % 10 + 9,
                    nb % 10 - nc % 10 + 9,
                    nc % 10 - nd % 10 + 9,
                )
                n = nd
                prev_price = n % 10
                for _ in range(3, 2000):
                    n = self.secret(n)
                    price = n % 10
                    a, b, c, d = b, c, d, prev_price - price + 9
                    prev_price = price
                    key = (a * 19**3) + (b * 19**2) + (c * 19) + d
                    if seen[key] == i:
                        continue
                    seen[key] = i
                    p[key] += price
            for k, v in p.items():
                m_ans[k] += v

        self.solve(seeds, worker)
        return max(m_ans)  # type:ignore[call-overload,no-any-return]

    @aoc_samples(
        (
            ("part_1", TEST1, 37327623),
            ("part_2", TEST2, 23),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 22)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_22.java
import static com.github.pareronia.aoc.IterTools.enumerate;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_22 extends SolutionBase<List<Integer>, Long, Integer> {

    private AoC2024_22(final boolean debug) {
        super(debug);
    }

    public static AoC2024_22 create() {
        return new AoC2024_22(false);
    }

    public static AoC2024_22 createDebug() {
        return new AoC2024_22(true);
    }

    @Override
    protected List<Integer> parseInput(final List<String> inputs) {
        return inputs.stream().map(Integer::valueOf).toList();
    }

    private int secret(final long seed) {
        long num = ((seed ^ (seed * 64)) % 16_777_216);
        num = (num ^ (num / 32)) % 16_777_216;
        return (int) (num ^ (num * 2048) % 16_777_216);
    }

    @Override
    public Long solvePart1(final List<Integer> seeds) {
        return seeds.stream()
                .mapToLong(
                        s -> {
                            long ss = s;
                            for (int i = 0; i < 2000; i++) {
                                ss = secret(ss);
                            }
                            return ss;
                        })
                .sum();
    }

    @Override
    public Integer solvePart2(final List<Integer> seeds) {
        final Map<Integer, Integer> p = new HashMap<>();
        final int[] seen = new int[19 * 19 * 19 * 19];
        Arrays.fill(seen, -1);
        enumerate(seeds.stream())
                .forEach(
                        e -> {
                            final int i = e.index();
                            int num = e.value();
                            final int na = num;
                            final int nb = secret(na);
                            final int nc = secret(nb);
                            final int nd = secret(nc);
                            int a;
                            int b = (9 + na % 10 - nb % 10);
                            int c = (9 + nb % 10 - nc % 10);
                            int d = (9 + nc % 10 - nd % 10);
                            num = nd;
                            int prevPrice = num % 10;
                            for (int j = 3; j < 2000; j++) {
                                num = secret(num);
                                final int price = num % 10;
                                a = b;
                                b = c;
                                c = d;
                                d = (9 + prevPrice - price);
                                prevPrice = price;
                                final int key = 6589 * a + 361 * b + 19 * c + d;
                                if (seen[key] == i) {
                                    continue;
                                }
                                seen[key] = i;
                                p.merge(key, price, Integer::sum);
                            }
                        });
        return p.values().stream().mapToInt(Integer::intValue).max().getAsInt();
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "37327623"),
        @Sample(method = "part2", input = TEST2, expected = "23")
    })
    public void samples() {}

    public static void main(final String[] args) throws Exception {
        AoC2024_22.create().run();
    }

    private static final String TEST1 =
            """
            1
            10
            100
            2024
            """;

    private static final String TEST2 =
            """
            1
            2
            3
            2024
            """;
}
